File: utils_python/stack.py
class StackNode:
  def __init__(self, data, next_node = None):
    self.data = data
    self.next_node = next_node

class Stack:
  def __init__(self):
    self.top = None

  def push(self, data):
    sn = StackNode(data, self.top)
    self.top = sn
    return True

  def pop(self):
    if self.top == None:
      raise IndexError

    value_to_return = self.top.data
    self.top = self.top.next_node
    return value_to_return

  def peek(self):
    if self.top == None:
      raise IndexError

    return self.top.data

  def is_empty(self):
    return self.top == None

def test():
  s = Stack()

  print('PASS' if s.is_empty() else 'FAIL')
  try:
    s.pop()
  except IndexError:
    pass

  try:
    s.peek()
  except IndexError:
    pass

  s.push(2)
  print('PASS' if s.peek() == 2 else 'FAIL')
  s.push(3)
  print('PASS' if s.peek() == 3 else 'FAIL')
  print('PASS' if s.pop() == 3 else 'FAIL')
  print('PASS' if s.pop() == 2 else 'FAIL')

  print('PASS' if s.is_empty() else 'FAIL')
  try:
    s.pop()
  except IndexError:
    pass

  try:
    s.peek()
  except IndexError:
    pass

if __name__ == '__main__':
  test()

File: utils_ruby/stack.rb
load "#{Dir.home}/Documents/Interviews/algo-practice/uni_node.rb"

class Stack
  attr_reader :length

  def initialize
    @top = nil
    @length = 0
  end

  def push(data)
    new_top = UniNode.new(data, @top)
    new_top.next_node = @top
    @top = new_top
    @length += 1
  end

  def pop
    raise "no item in the stack" if @top.nil?
    data_to_pop = @top.data
    @top = @top.next_node
    @length -= 1
    return data_to_pop
  end

  def is_empty?
    return @top.nil?
  end

  def peek
    raise "no item in the stack" if @top.nil?
    return @top.data
  end

  def inspect
    msg = ''
    node = @top
    while !node.nil? do
      msg << "#{node.data}" << "->"
      node = node.next_node
    end
    msg << "nil"
    msg
  end
end

def test
  s = Stack.new
  begin
    puts "PEEK empty stack"
    s.peek
  rescue StandardError => e
    puts e.message
    puts
  end

  begin
    puts "POP empty stack"
    s.pop
  rescue StandardError => e
    puts e.message
    puts
  end

  puts "ADD 3, 5, 6"
  s.push(3)
  s.push(5)
  s.push(6)
  puts s.length == 3 ? 'PASS' : 'FAIL'
  puts s.inspect
  puts

  puts "POP and PEEK"
  puts s.pop == 6 ? 'PASS' : 'FAIL'
  puts s.peek == 5 ? 'PASS' : 'FAIL'
  puts s.length == 2 ? 'PASS' : 'FAIL'
  puts s.inspect
  puts
end
